R/model_componets.R

```diff
@@ -174,68 +177,118 @@ carbon_fluxes_internal <- function(env, state = NULL, params = NULL){
+carbon_fluxes <-
+  function(DOMdecomp = "MM",
+           POMdecomp = "MM",
+           MBdecay = "LM",
+           env = NULL,
+           state = NULL,
+           params = NULL) {
+    # Check the inputs
+    assert_that(any(is.null(env) | is.environment(env)))
+    assert_that(any(is.null(state) | is.list(state)))
+    assert_that(any(is.null(state) | is.data.frame(params)))
+    assert_that(all(sapply(list(POMdecomp, DOMdecomp, MBdecay), is.character)))
+    assert_that(sum(DOMdecomp %in% c("MM", "RMM", "ECA", "LM")) == 1, msg = 'DOMdecomp must be "MM", "RMM", "ECA", "LM"')
+    assert_that(sum(POMdecomp %in% c("MM", "RMM", "ECA", "LM")) == 1, msg = 'POMdecomp must be "MM", "RMM", "ECA", "LM"')
+    assert_that(sum(MBdecay %in% c("MM", "RMM", "ECA", "LM")) == 1, msg = 'MBdecay must be "MM", "RMM", "ECA", "LM"')
+
+    # Create the list to store the output.
+    out <- list()
+
+    # Store information about the fluxes being used.
+    out[["flux_table"]] <- data.table::data.table("DOMdecomp" = DOMdecomp,
+                                                  "POMdecomp" = POMdecomp,
+                                                  "MBdecay" = MBdecay)
+
+    # Store the fluxes.
+    out[["flux_function"]] <- function(env,
+                                       state = NULL,
+                                       params = NULL) {
+      # Update model parameter & initial state values if needed.
+      if (!is.null(params)) {
+        p <- params$value
+        names(p) <- params$parameter
+        modify_env(env, p)
 
-  assert_that(is.character(POMdecomp))
-  assert_that(length(POMdecomp) == 1)
-  assert_that(any(POMdecomp %in% c("MM", "RMM", "ECA", "LM")))
-
-  # Create the list to store the output.
-  out <- list()
+      }
+      if (!is.null(state)) {
+        modify_env(env, state)
+      }
+      with(as.list(env), {
+        # Set up the original functions.
+        fluxes  <- carbon_fluxes_internal(env = env)
+
+        # If else statement determining the kinetics used in the decompostion of the DOM.
+        if (DOMdecomp == "MM") {
+          fluxes[["F1"]] = function() {
+            # DOC uptake by microbial biomass.
+            # Michaelis menten kinetics
+            (1 / E.c) * (V.d + m.r) * B * D / (K.d + D)
+          }
+        } else if (DOMdecomp == "RMM") {
+          fluxes[["F1"]] = function() {
+            # DOC uptake by microbial biomass.
+            # Reverse michaelis menten kinetics
+            (1 / E.c) * (V.d + m.r) * B * D / (K.d + B)
+          }
+        } else if (DOMdecomp == "ECA") {
+          fluxes[["F1"]] = function() {
+            # DOC uptake by microbial biomass, note that this uses ECA kinetics.
+            (1 / E.c) * (V.d + m.r) * B * D / (K.d + B + D)
+          }
+        } else if (DOMdecomp == "LM") {
+          fluxes[["F1"]] = function() {
+            # DOC uptake by microbial biomass, note that this uses a linear model of kinetics.
+            V.d * B * D / (K.d + B)
```

Comment:
Is this correct for the linear model? or should it look more along the lines of 

`(1 / E.c) * (V.d + m.r) * B * D / (K.d + B)` ?